### PR TITLE
ARCHIE-206 - Add support for gpt-3.5-turbo-16k

### PR DESCRIPTION
**Pull request by cgianelle**
https://cipherhealth.atlassian.net/browse/ARCHIE-206

Also addresses:
https://cipherhealth.atlassian.net/browse/ARCHIE-178 (Dynamic token counts)

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    func(message)
+
+'''
+gpt-3.5-turbo-16k offers 4 times the context length of gpt-3.5-turbo at twice the price: 
+$0.003 per 1K input tokens and $0.004 per 1K output tokens. 
+
+So if the conversation exceeds the cutoff, then switch to using the extended model. Otherwise, use 
+the standard model, as that seems fine for most conversations at this point.
+'''
+def determine_openai_model_to_use(input_token_count):
```
Comment (chuchothecat) [inline main.py:73]: 👍

**Review by chuchothecat — APPROVED** [review]
LGTM, nicely done Chris!

File: slack_gpt_bot.py [hunk 2]
```diff
@@ -63,8 +65,41 @@ def logging_wrapper(message, severity=logging.INFO, **kwargs):
     func = log.get(severity, json_std_logger.info)
     func(message)
 
+'''
+gpt-3.5-turbo-16k offers 4 times the context length of gpt-3.5-turbo at twice the price: 
+$0.003 per 1K input tokens and $0.004 per 1K output tokens. 
+
+So if the conversation exceeds the cutoff, then switch to using the extended model. Otherwise, use 
+the standard model, as that seems fine for most conversations at this point.
```
Comment (markusos) [inline slack_gpt_bot.py:73]: How does it work if it switches model mid conversation? Does it lose all the context or is that moved over as well?

File: slack_gpt_bot.py [hunk 3]
```diff
@@ -82,14 +117,15 @@ def command_handler(body, context):
...
         didn't originate from this channel, you got a polite message and 
         were denied access.
+        
+        if channel_id != 'C057NBLL2G4': #lock to test channel for beta
+            slack_resp = app.client.chat_postMessage( 
+                channel=channel_id,
+                thread_ts=thread_ts,
+                text="Our apologies, however the Beta ChatGPT bot is not allowed outside of the beta-slack-chatgpt-bot channel"
+            )
+            return
```
Comment (markusos) [inline slack_gpt_bot.py:127]: Do we want to add the channel lock back again?

**Review by markusos — APPROVED** [review]
Looks good, but added some questions

Would also be nice to run the code through a code linter and formatter. We use https://pypi.org/project/flake8/ and https://pypi.org/project/black/ for our python projects on the data team.

File: slack_gpt_bot.py (hunk 2 above)
Comment (cgianelle) [reply slack_gpt_bot.py:73]: the bot works by using the Slack conversation thread. 

https://github.com/SolarCS/slack-gpt-bot/blob/621cf233e026ffba28c7c3897ec88a0020133241/slack_gpt_bot.py#L137C1-L139

So it processes the entire conversation in that thread as one message, and sends that to open AI. That's how the bot preserves the context. The difference between the 2 models should only be the token count and cost. Most conversations are brief, so this is a way to extend the conversation without overpaying for the short ones. Or at least that is the idea

File: slack_gpt_bot.py (hunk 3 above)
Comment (cgianelle) [reply slack_gpt_bot.py:127]: No, that was for our beta release back in May. I should remove the comment as it is preserved by Github if we ever needed to know what we did 😄

**Pull request merged**